README
pdlua -- a Lua embedding for Pd
Copyright (C) 2007,2008,2009,2013 Claude Heiland-Allen <claude@mathr.co.uk>
Copyright (C) 2012 Martin Peach martin.peach@sympatico.ca


This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.


Current maintainer: Albert Gräf <aggraef@gmail.com>
This source is available on GitHub: https://github.com/agraef/pd-lua

Since the contents of the original README has become rather outdated, I've
updated the information below.

pdlua works with all Pd flavors out there and makes Pd external programming
incredibly easy (please check the included examples and the tutorial in the
corresponding subdirectories). Originally written by Claude Heiland-Allen,
pdlua has gone through the hands of a few people over the years, including
mrpeach (maintainer since 2011), zmoelnig a.k.a. umlaeute (loader update,
Debian package), and myself (Arch package, Lua 5.3+ support, Purr Data and
plugdata support, tutorial). Please also check my brief account on the history
of pd-lua below.

Lua 5.4 is highly recommended with the latest version, Lua 5.3 works as
well. Reportedly, Lua 5.2 and even 5.1 still work (at least to some extent),
but I haven't tested these myself for quite a while, so you may run into
problems with these old versions - use at your own risk.

If you haven't used pdlua before, please make sure to check the included
tutorial first. See:

https://agraef.github.io/pd-lua/tutorial/pd-lua-intro.html

This contains a fairly gentle introduction to pdlua, walks you through the
creation of a basic example, and then goes on to cover all major features of
pdlua, including the special facilities for tables, clocks, receivers, and
live-coding, in quite some detail. This will be helpful when embarking on your
own projects, or trying to make sense of the included examples.


History and Credits:

pdlua was originally written by Claude Heiland-Allen, with contributions by
Frank Barknecht and Martin Peach, according to the original source.

Claude's repository is still online at https://code.mathr.co.uk/pdlua; please
also check Claude's website at https://mathr.co.uk/, and
https://mathr.co.uk/blog/lua.html for Lua-related content on his blog
(including pdlua's original announcement).

Martin Peach took over maintenance of pdlua in 2011, you can find that part of
the history in Pd's old svn repository at SourceForge, including IOhannes
Zmölnig's loader update for Pd 0.47. See
https://sourceforge.net/p/pure-data/svn/HEAD/tree/trunk/externals/loaders/pdlua/

The present source is a fork of umlaeute's repository at
https://anonscm.debian.org/git/pkg-multimedia/pd-lua.git, from which the
official Debian packages are built. This in turn is apparently based on the
0.7.3 version of pdlua in Pd's svn repository.

The current version (0.8 and later) is actively maintained by Albert Gräf <aggraef@gmail.com>, with contributions by Claude Heiland-Allen (Lua 5.4 update, bugfixes), Alexandre Porres (documentation updates), sebshader (relative path loading), and Timothy Schoen (plugdata compatibility). This is also the version included in Purr Data (https://github.com/agraef/purr-data) and Timothy Schoen's plugdata (https://github.com/plugdata-team/plugdata).

Please file bugs and requests at https://github.com/agraef/pd-lua. I maintain
a lot of packages and thus it may take a while at times, but I look at all the
bug reports and pull requests, and try to respond to them eventually. :)


Ready-Made Packages:

The latest vanilla binaries for Ubuntu, Mac and Windows can always be found on
https://github.com/agraef/pd-lua. The same version is also in the Arch
community repository (maintained by dvzrv, thanks David!). On other Linux
systems you'll have to compile pdlua yourself from source, which isn't hard to
do (see below). Also, both Purr Data and plugdata ship with pdlua, so no 3rd
party package is needed.

To enable the pdlua loader in Pd after installation, just add `pdlua` to your
startup libraries (after adding its parent directory to Pd's search path if
necessary) and you should be set.

Mac users please note that the packages I distribute aren't notarized, so on
recent macOS versions you'll have to jump through the usual hoops to make them
usable. Running `xattr -rd com.apple.quarantine` on pdlua.pd_darwin should
usually do the trick, but check the internet for up-to-date information on
Gatekeeper for details.


Compilation Instructions:

The source should compile out of the box on (at least) Linux, macOS and
Windows, just make sure that you have Lua and Pd installed and run
`make`. pdlua uses pd-lib-builder by Katja Vetter as its build system, please
check https://github.com/pure-data/pd-lib-builder and the included
Makefile.pdlibbuilder for details.

pdlua works with any reasonably modern Pd flavour, but pd-lib-builder will
look for vanilla Pd (http://msp.ucsd.edu/software.html). So you'll need to
have that installed, or be prepared to adjust the Makefile accordingly.

If you compile from git sources and use the included lua submodule (see
Compilation from Git Sources below), then only Pd will be needed. Otherwise
you'll also have to install Lua (https://www.lua.org).

The following build environments on various platforms are known to work and
are supported.

Linux: Uses gcc and GNU make. Pd and Lua should be available in your package
repositories. If not then it's easy to install them from source.

Mac: Uses Xcode. The build requires that you have the Pd header files
somewhere under /Applications/Pd-*/Contents/Resources/src (which should be the
case if you're running a recent Pd distribution by MSP); otherwise you can set
the `PDINCLUDEDIR` make variable accordingly. Lua is available, e.g., in
Homebrew (https://brew.sh/).

Windows: Uses msys2/mingw (https://www.msys2.org/), which is the Windows
incarnation of gcc and GNU make. The build assumes that you have a recent Pd
installation as distributed by MSP, so that the header files and pd.dll are
found during compilation and linkage, respectively. Lua is available in msys2.


Compilation from Git Sources:

The repository source includes a lua submodule which can be used to compile
Lua alongside pdlua, so no Lua installation is needed to compile the
external. (You'll still need to have Pd installed.) Just make sure to check
out the submodule with the repository (`git clone --recurse-submodules`).
If you already checked out the repository, you can also do `git submodule
update --init` instead. This should get you a fairly recent revision of the
upstream Lua source.

Once the lua submodule is checked out, you can just run `make` as usual, it
will use the submodule to compile Lua alongside the pdlua module. The
resulting binary will be statically linked with the Lua interpreter.

Also note that on the Mac you can build a universal (a.k.a. arm64+x86_64)
binary this way, using `arch="arm64 x86_64"` when running `make`. This will
work best if Lua is compiled alongside pdlua, since installed Lua libraries
may not be universal binaries (in Homebrew at least they aren't).

Otherwise, if the lua submodule isn't checked out, `make` will try
`pkg-config` to locate an existing Lua installation on your system and link
against that.


Installation:

After a successful compile, you can go about installing the external with the
usual `make install` which copies the external to its own directory named
pdlua. This should generally do something sensible on each of the supported
platforms. Note that the installation may require root privileges.

Linux: By default, installation goes into /usr/local/lib/pd-externals, which
should be one of the directories on Pd's built-in search path.

Mac: Installation goes into ~/Library/Pd by default. Move the pdlua folder to
/Library/Pd for a system-wide installation.

Windows/mingw: Installation goes into the Pd subdirectory of your %APPDATA%
folder by default. Move the pdlua folder to the extra folder of Pd in your
ProgramFiles directory for a system-wide installation.

In any case, it's possible to override the default installation directory
using the PDLIBDIR variable. E.g., on Linux you can install into Pd's
/usr/lib/pd/extra folder as follows:

    make install PDLIBDIR=/usr/lib/pd/extra

It's generally advisable to do a staged install using `DESTDIR` first so that
you can review the installation directory layout beforehand and adjust it if
needed. To these ends, run `make install`, e.g., as follows:

    make install DESTDIR=./build

It is possible to use DESTDIR along with PDLIBDIR.

Another possibility is to run `make install objectsdir=./build` which creates
the `pdlua` folder directly in the `build` directory from where you can move
it to any location on Pd's library search path that you see fit. (This option
is considered deprecated, but it's still the most convenient for a quick
manual installation.)


debian/changelog
pd-lua (0.11.6+ds1-1) unstable; urgency=medium

  * New upstream version 0.11.6+ds1
    + Drop obsolete patches
    + Switch buildsystem to pd-lib-builder
    + Add patch for cross-building
    + B-D on liblua5.4-dev
  * Build for both single and double-precision Pd
  * Repack upstream sources
    + Exclude repository config from source tarball
    + Exclude generated documentation
    + Add '+ds' repacksuffix
  * Update upstream information
    + Update upstream contact
    + Watch new upstream repository
  * Update copyright information
    + Update d/copyright
    + Exclude binary files from 'licensecheck'
    + Re-generate d/copyright_hints
  * Apply 'wrap-and-sort -ast'
  * Bump standards version to 4.6.2

 -- IOhannes m zmölnig (Debian/GNU) <umlaeute@debian.org>  Wed, 05 Jul 2023 15:02:17 +0200

pd-lua (0.7.3-6) unstable; urgency=medium

  * Modernize 'licensecheck' target
    + Ensure that 'licensecheck' is run with the C.UTF-8 locale
    + Exclude debian/ from 'licensecheck'
    + Re-generate d/copyright_hints

 -- IOhannes m zmölnig (Debian/GNU) <umlaeute@debian.org>  Wed, 07 Dec 2022 10:47:01 +0100

pd-lua (0.7.3-5) unstable; urgency=medium

  * Apply "warp-and-sort -ast"
  * Bump standards version to 4.6.1

 -- IOhannes m zmölnig (Debian/GNU) <umlaeute@debian.org>  Sun, 04 Sep 2022 13:55:19 +0200

pd-lua (0.7.3-4) unstable; urgency=medium

  [ Debian Janitor ]
  * Set upstream metadata fields: Archive, Bug-Database, Bug-Submit,
    Repository, Repository-Browse.

  [ IOhannes m zmölnig (Debian/GNU) ]
  * Add salsa-ci configuration
  * Bump dh-compat to 13
  * Bump standards version to 4.5.0

 -- IOhannes m zmölnig (Debian/GNU) <umlaeute@debian.org>  Wed, 19 Aug 2020 21:48:05 +0200

pd-lua (0.7.3-3) unstable; urgency=medium

  [ Ondřej Nový ]
  * Use debhelper-compat instead of debian/compat

  [ IOhannes m zmölnig ]
  * Fix FTCBFS
    + Supply pkg-config from dpkg's buildtools.mk
    + Don't strip during make install.
    Thanks to Helmut Grohne <helmut@subdivi.de> (Closes: #930944)
  * Declare that building this package doesn't require "root" powers
  * Remove obsolete d/source/local-options
  * Bump dh compat to 12
  * Bump standards version to 4.4.0

 -- IOhannes m zmölnig (Debian/GNU) <umlaeute@debian.org>  Wed, 11 Sep 2019 15:40:38 +0200

pd-lua (0.7.3-2) unstable; urgency=medium

  * Switched buildsystem from dh to cdbs
    * Bumped dh compat to 11
    * Enabled hardening
  * Fixed spelling of 'Lua'
  * Updated Vcs-* stanzas to salsa.d.o
  * Updated maintainer address
  * Switched URLs to https://
  * Removed obsolete d/git-tuneclone.sh
    * Dropped d/README.source from docs
  * Updated d/copyright_hints
  * Bumped standards version to 4.1.3

 -- IOhannes m zmölnig (Debian/GNU) <umlaeute@debian.org>  Thu, 01 Feb 2018 23:19:43 +0100

pd-lua (0.7.3-1) unstable; urgency=medium

  * Imported Upstream version 0.7.3

  * Removed patches applied upstream
  * Backported patch to fix format-security errors
  * Updated files to install
  * Updated build-flags to new upstream build-system
  * Override build-date for reproducible builds
  * Depend on 'puredata' rather than 'puredata-core'
  * Update Vcs-* stanzas
  * Added debian/git-tuneclone.sh script (see debian/README.source)
  * Regenerated debian/(control|copyright_hints)
  * Updated debian/copyright
  * Bumped Standards-version to 3.9.6

 -- IOhannes m zmölnig (Debian/GNU) <umlaeute@debian.org>  Sun, 24 Jan 2016 22:25:22 +0100

pd-lua (0.6.0-2) unstable; urgency=medium

  * Added "-fPIC" to build-flags (Closes: #759830)
  * Fixed installation path for examples/
  * Document that ignoring debian/control.in is no show-stopper

 -- IOhannes m zmölnig (Debian/GNU) <umlaeute@debian.org>  Mon, 01 Sep 2014 09:32:15 +0200

pd-lua (0.6.0-1) unstable; urgency=medium

  * Initial release. (Closes: #755268)

 -- IOhannes m zmölnig (Debian/GNU) <umlaeute@debian.org>  Mon, 21 Jul 2014 18:13:24 +0200


pd.lua
--[[
pdlua -- a Lua embedding for Pd
Copyright (C) 2007,2008 Claude Heiland-Allen <claude@mathr.co.uk>
Copyright (C) 2012 Martin Peach martin.peach@sympatico.ca

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
--]]

-- storage for Pd C<->Lua interaction
pd._classes = { } -- take absolute paths and turn them into classes
pd._pathnames = { } -- look up absolute path by creation name
pd._objects = { }
pd._clocks = { }
pd._receives = { }
pd._loadpath = ""

-- add a path to Lua's "require" search paths
pd._setrequirepath = function(path)
  pd._packagepath = package.path
  pd._packagecpath = package.cpath
  if (pd._iswindows) then
    package.path = path .. "\\?;" .. path .. "\\?.lua;" .. package.path
    package.cpath = path .. "\\?.dll;" .. package.cpath
  else
    package.path = path .. "/?;" .. path .. "/?.lua;" .. package.path
    package.cpath = path .. "/?.so;" .. package.cpath
  end
end

-- reset Lua's "require" search paths
pd._clearrequirepath = function()
  package.path = pd._packagepath
  package.cpath = pd._packagecpath
end

-- check if we need to register a basename class first
pd._checkbase = function (name)
  return pd._pathnames[name] == true
end

-- constructor dispatcher
pd._constructor = function (name, atoms)
  local fullpath = pd._pathnames[name]
  if nil ~= pd._classes[fullpath] then
    local o = pd._classes[fullpath]:new():construct(name, atoms)
    if o then
      pd._objects[o._object] = o
      return o._object
    end
  end
  return nil
end

-- destructor dispatcher
pd._destructor = function (object)
  if nil ~= pd._objects[object] then
    pd._objects[object]:destruct()
  end
end

-- inlet method dispatcher
pd._dispatcher = function (object, inlet, sel, atoms)
  if nil ~= pd._objects[object] then
    pd._objects[object]:dispatch(inlet, sel, atoms)
  end
end

-- clock method dispatcher
pd._clockdispatch = function (c)
  if nil ~= pd._clocks[c] then
    local m = pd._clocks[c]._method
    pd._clocks[c]._target[m](pd._clocks[c]._target)
  end
end

--whoami method dispatcher
pd._whoami = function (object)
  if nil ~= pd._objects[object] then
    return pd._objects[object]:whoami()
  end
end

--class method dispatcher
pd._get_class = function (object)
  if nil ~= pd._objects[object] then
    return pd._objects[object]:get_class()
  end
end

-- prototypical OO system
pd.Prototype = { }

function pd.Prototype:new(o)
  o = o or {}
  setmetatable(o, self)
  self.__index = self
  return o
end

-- clocks
pd.Clock = pd.Prototype:new()

function pd.Clock:register(object, method)
  if nil ~= object then
    if nil ~= object._object then
      self._clock = pd._createclock(object._object, method)
      self._target = object
      self._method = method
      pd._clocks[self._clock] = self
      return self
    end
  end
  return nil
end

function pd.Clock:destruct()
  pd._clocks[self._clock] = nil
  pd._clockfree(self._clock)
  self._clock = nil
end

function pd.Clock:dispatch()
  local m = self._target[self._method]
  if type(m) == "function" then
    return m(self._target)
  else
    self._target:error(
      "no method for `" .. self._method ..
      "' at clock of Lua object `" .. self._name .. "'"
    )
  end
end

function pd.Clock:set(systime)
  pd._clockset(self._clock, systime)
end

function pd.Clock:delay(delaytime)
  pd._clockdelay(self._clock, delaytime)
end

function pd.Clock:unset()
  pd._clockunset(self._clock)
end

-- tables
pd.Table = pd.Prototype:new()

function pd.Table:sync(name)
  self.name = name
  self._length, self._array = pd._getarray(name)
  if self._length < 0 then
    return nil
  else
    return self
  end
end

function pd.Table:destruct()
  self._length = -3
  self._array = nil
end

function pd.Table:get(i)
  if type(i) == "number" and 0 <= i and i < self._length then
    return pd._readarray(self._length, self._array, i)
  else
    return nil
  end
end

function pd.Table:set(i, f)
  if type(i) == "number" and type(f) == "number" and 0 <= i and i < self._length then
    return pd._writearray(self._length, self._array, i, f)
  else
    return nil
  end
end

function pd.Table:length()
  if self._length >= 0 then
    return self._length
  else
    return nil
  end
end

function pd.Table:redraw()
  pd._redrawarray(self.name)
end

-- receivers
function pd._receivedispatch(receive, sel, atoms)
  if nil ~= pd._receives[receive] then
    pd._receives[receive]:dispatch(sel, atoms)
  end
end

pd.Receive = pd.Prototype:new()

function pd.Receive:register(object, name, method)
  if nil ~= object then
    if nil ~= object._object then
      self._receive = pd._createreceive(object._object, name)
      self._name = name
      self._target = object
      self._method = method
      pd._receives[self._receive] = self
      return self
    end
  end
  return nil
end

function pd.Receive:destruct()
  pd._receives[self._receive] = nil
  pd._receivefree(self._receive)
  self._receive = nil
  self._name = nil
  self._target = nil
  self._method = nil
end

function pd.Receive:dispatch(sel, atoms)
  self._target[self._method](self._target, sel, atoms)
end

-- patchable objects
pd.Class = pd.Prototype:new()

function pd.Class:register(name)
  -- if already registered, return existing
  local regname
  local fullpath = pd._loadpath .. name
  if nil ~= pd._classes[fullpath] then
    return pd._classes[fullpath]
  end
  if pd._loadname then
    -- don't alter existing classes of basename,
    -- if another file has ownership of basename
    if not pd._pathnames[name] then
      pd._pathnames[name] = true
    end
    regname = pd._loadname
  else
    regname = name
  end
  pd._pathnames[regname] = fullpath
  pd._classes[fullpath] = self       -- record registration
  self._class = pd._register(name)  -- register new class
  self._name = name
  self._loadpath = pd._loadpath
  if name == "pdlua" then
    self._scriptname = "pd.lua"
  else
    self._scriptname = name .. ".pd_lua"
  end -- mrpeach 20111027
  return self                       -- return new
end

function pd.Class:construct(sel, atoms)
  self._object = pd._create(self._class)
  self.inlets = 0
  self.outlets = 0
  self._canvaspath = pd._canvaspath(self._object)
  if self:initialize(sel, atoms) then
    pd._createinlets(self._object, self.inlets)
    pd._createoutlets(self._object, self.outlets)
    self:postinitialize()
    return self
  else
    return nil
  end
end

function pd.Class:destruct()
  pd._objects[self] = nil
  self:finalize()
  pd._destroy(self._object)
end

function pd.Class:dispatch(inlet, sel, atoms)
  local m = self[string.format("in_%d_%s", inlet, sel)]
  if type(m) == "function" then
    if sel == "bang"    then return m(self)           end
    if sel == "float"   then return m(self, atoms[1]) end
    if sel == "symbol"  then return m(self, atoms[1]) end
    if sel == "pointer" then return m(self, atoms[1]) end
    if sel == "list"    then return m(self, atoms)    end
    return m(self, atoms)
  end
  m = self["in_n_" .. sel]
  if type(m) == "function" then
    if sel == "bang"    then return m(self, inlet)           end
    if sel == "float"   then return m(self, inlet, atoms[1]) end
    if sel == "symbol"  then return m(self, inlet, atoms[1]) end
    if sel == "pointer" then return m(self, inlet, atoms[1]) end
    if sel == "list"    then return m(self, inlet, atoms)    end
    return m(self, inlet, atoms)
  end
  m = self[string.format("in_%d", inlet)]
  if type(m) == "function" then
    return m(self, sel, atoms)
  end
  m = self["in_n"]
  if type(m) == "function" then
    return m(self, inlet, sel, atoms)
  end
  self:error(
     string.format("no method for `%s' at inlet %d of Lua object `%s'",
		   sel, inlet, self._name)
  )
end

function pd.Class:outlet(outlet, sel, atoms)
  pd._outlet(self._object, outlet, sel, atoms)
end

function pd.Class:initialize(sel, atoms) end

function pd.Class:postinitialize() end

function pd.Class:finalize() end

function pd.Class:dofilex(file)
  -- in case of register being called, make sure
  -- classes in other paths aren't getting affected
  -- save old loadname in case of weird nesting loading
  local namesave = pd._loadname
  local pathsave = pd._loadpath
  pd._loadname = nil
  pd._loadpath = self._loadpath
  local f, path = pd._dofilex(self._class, file)
  pd._loadname = namesave
  pd._loadpath = pathsave
  return f, path
end

function pd.Class:dofile(file)
  -- in case of register being called, make sure
  -- classes in other paths aren't getting affected
  -- save old loadname in case of weird nesting loading
  local namesave = pd._loadname
  local pathsave = pd._loadpath
  pd._loadname = nil
  pd._loadpath = self._loadpath
  local f, path = pd._dofile(self._object, file)
  pd._loadname = namesave
  pd._loadpath = pathsave
  return f, path
end

function pd.Class:error(msg)
  pd._error(self._object, msg)
end

function pd.Class:whoami()
  return self._scriptname or self._name
end

function pd.Class:get_class() -- accessor for t_class*
  return self._class or nil
end

local lua = pd.Class:new():register("pdlua")  -- global controls (the [pdlua] object only)

function lua:initialize(sel, atoms)
  self.inlets = 1
  self.outlets = 0    -- FIXME: might be nice to have errors go here?
  return true
end

function lua:in_1_load(atoms)  -- execute a script
  self:dofile(atoms[1])
end


local luax = pd.Class:new():register("pdluax")  -- classless lua externals (like [pdluax foo])

function luax:initialize(sel, atoms)          -- motivation: pd-list 2007-09-23
  if not atoms[1] then
    -- create a dummy object, which can still be clicked for help
    self.inlets = 0
    self.outlets = 0
    self._scriptname = ""
    return true
  end
  local f, pathname = self:dofile(atoms[1] .. ".pd_luax")
  if f and pathname then
    local function basename(str)
      return string.gsub(str, "(.*/)(.*)", "%2")
    end
    self._scriptname = pathname .. '/' .. basename(atoms[1]) .. ".pd_luax" -- mrpeach 20120201
    local atomstail = { }          -- munge for better lua<->luax compatibility
    for i,_ in ipairs(atoms) do                  
      if i > 1 then
        atomstail[i-1] = atoms[i]
      end
    end
    return f(self, atoms[1], atomstail)
  else
    return false   -- error message already output by dofile()
  end
end

-- fin pd.lua
